Error while deploying integration in Knative based environment

I am trying to deploy the following integration called `Greeter.java` with content

```java
import org.apache.camel.builder.RouteBuilder;

public class Greeter extends RouteBuilder {
    @Override
    public void configure() throws Exception {

        restConfiguration().component("undertow").port(8080);

        rest()
                .post("/greeter")
                .route()
                .log("${body}");

    }
}
```

When i deploy the service using the command `kamel run --name=greeter Greeter.java` or  `kamel run   Greeter.java`  it throws an error 
```
Exception in thread "main" org.joor.ReflectException: Compilation error: /CAMEL_K_ROUTE_000.java:4: error: class Greeter is public, should be declared in a file named Greeter.java
public class Greeter extends RouteBuilder {
       ^
1 error

    at org.joor.Compile.compile(Compile.java:97)
    at org.joor.Reflect.compile(Reflect.java:102)
    at org.joor.Reflect.compile(Reflect.java:77)
    at org.apache.camel.k.jvm.loader.JavaSourceLoader$1.configure(JavaSourceLoader.java:39)
    at org.apache.camel.builder.RouteBuilder.checkInitialized(RouteBuilder.java:463)
    at org.apache.camel.builder.RouteBuilder.configureRoutes(RouteBuilder.java:403)
    at org.apache.camel.builder.RouteBuilder.addRoutesToCamelContext(RouteBuilder.java:384)
    at org.apache.camel.impl.DefaultCamelContext$3.call(DefaultCamelContext.java:1024)
    at org.apache.camel.impl.DefaultCamelContext$3.call(DefaultCamelContext.java:1021)
    at org.apache.camel.impl.DefaultCamelContext.doWithDefinedClassLoader(DefaultCamelContext.java:3267)
    at org.apache.camel.impl.DefaultCamelContext.addRoutes(DefaultCamelContext.java:1021)
    at org.apache.camel.main.MainSupport.postProcessCamelContext(MainSupport.java:612)
    at org.apache.camel.main.MainSupport.postProcessContext(MainSupport.java:550)
    at org.apache.camel.k.jvm.Runtime.doStart(Runtime.java:107)
    at org.apache.camel.support.ServiceSupport.start(ServiceSupport.java:72)
    at org.apache.camel.main.MainSupport.run(MainSupport.java:170)
    at org.apache.camel.k.jvm.Application.main(Application.java:73)
```

--- a/pkg/trait/knative_service.go
+++ b/pkg/trait/knative_service.go
@@ -140,6 +140,8 @@ func (t *knativeServiceTrait) getServiceFor(e *Environment) (*serving.Service, e
 		envvar.SetVal(&environment, envName, s.Content)
 
 		params := make([]string, 0)
+		params = append(params, "name="+s.Name)
+
 		if s.InferLanguage() != "" {
 			params = append(params, "language="+string(s.InferLanguage()))
 		}

--- a/pkg/trait/knative_test.go
+++ b/pkg/trait/knative_test.go
@@ -111,7 +111,7 @@ func TestKnativeTraitWithCompressedSources(t *testing.T) {
 
 		routes := util.LookupEnvVar(vars, "CAMEL_K_ROUTES")
 		assert.NotNil(t, routes)
-		assert.Equal(t, "env:CAMEL_K_ROUTE_000?language=js&compression=true", routes.Value)
+		assert.Equal(t, "env:CAMEL_K_ROUTE_000?name=routes.js&language=js&compression=true", routes.Value)
 
 		route := util.LookupEnvVar(vars, "CAMEL_K_ROUTE_000")
 		assert.NotNil(t, route)
@@ -209,7 +209,7 @@ func TestKnativeTraitWithConfigMapSources(t *testing.T) {
 
 		routes := util.LookupEnvVar(vars, "CAMEL_K_ROUTES")
 		assert.NotNil(t, routes)
-		assert.Equal(t, "env:CAMEL_K_ROUTE_000?language=js&compression=true", routes.Value)
+		assert.Equal(t, "env:CAMEL_K_ROUTE_000?name=routes.js&language=js&compression=true", routes.Value)
 
 		route := util.LookupEnvVar(vars, "CAMEL_K_ROUTE_000")
 		assert.NotNil(t, route)

--- a/runtime/camel-k-runtime-core/src/main/java/org/apache/camel/k/Source.java
+++ b/runtime/camel-k-runtime-core/src/main/java/org/apache/camel/k/Source.java
@@ -23,14 +23,20 @@ import org.apache.camel.util.URISupport;
 import org.apache.commons.lang3.StringUtils;
 
 public class Source {
+    private final String name;
     private final String location;
     private final String language;
     private final boolean compressed;
 
-    private Source(String location, String language, boolean compression) {
+    private Source(String name, String location, String language, boolean compression) {
+        this.name = name;
         this.location = location;
         this.language = language;
         this.compressed = compression;
+    }
+
+    public String getName() {
+        return name;
     }
 
     public String getLocation() {
@@ -66,7 +72,8 @@ public class Source {
         final String query = StringUtils.substringAfter(uri, "?");
         final Map<String, Object> params = URISupport.parseQuery(query);
         final String languageName = (String) params.get("language");
-        final boolean compression = Boolean.valueOf((String) params.get("compression"));
+        final String compression = (String) params.get("compression");
+
 
         String language = languageName;
         if (ObjectHelper.isEmpty(language)) {
@@ -77,7 +84,16 @@ public class Source {
             throw new IllegalArgumentException("Unknown language " + language);
         }
 
+        String name = (String) params.get("name");
+        if (name == null) {
+            name = StringUtils.substringAfter(location, ":");
+            name = StringUtils.substringBeforeLast(name, ".");
 
-        return new Source(location, language, compression);
+            if (name.contains("/")) {
+                name = StringUtils.substringAfterLast(name, "/");
+            }
+        }
+
+        return new Source(name, location, language, Boolean.valueOf(compression));
     }
 }

--- a/runtime/camel-k-runtime-jvm/src/main/java/org/apache/camel/k/jvm/loader/JavaClassLoader.java
+++ b/runtime/camel-k-runtime-jvm/src/main/java/org/apache/camel/k/jvm/loader/JavaClassLoader.java
@@ -20,7 +20,6 @@ import java.util.Collections;
 import java.util.List;
 
 import org.apache.camel.builder.RouteBuilder;
-import org.apache.camel.k.Constants;
 import org.apache.camel.k.RoutesLoader;
 import org.apache.camel.k.Runtime;
 import org.apache.camel.k.Source;
@@ -34,14 +33,13 @@ public class JavaClassLoader implements RoutesLoader {
 
     @Override
     public RouteBuilder load(Runtime.Registry registry, Source source) throws Exception {
-        String path = source.getLocation();
-        path = StringUtils.removeStart(path, Constants.SCHEME_CLASSPATH);
-        path = StringUtils.removeEnd(path, ".class");
+        String name = source.getName();
+        name = StringUtils.removeEnd(name, ".class");
 
-        Class<?> type = Class.forName(path);
+        Class<?> type = Class.forName(name);
 
         if (!RouteBuilder.class.isAssignableFrom(type)) {
-            throw new IllegalStateException("The class provided (" + path + ") is not a org.apache.camel.builder.RouteBuilder");
+            throw new IllegalStateException("The class provided (" + source.getLocation() + ") is not a org.apache.camel.builder.RouteBuilder");
         }
 
         return (RouteBuilder)type.newInstance();

--- a/runtime/camel-k-runtime-jvm/src/main/java/org/apache/camel/k/jvm/loader/JavaSourceLoader.java
+++ b/runtime/camel-k-runtime-jvm/src/main/java/org/apache/camel/k/jvm/loader/JavaSourceLoader.java
@@ -49,15 +49,15 @@ public class JavaSourceLoader implements RoutesLoader {
                 final CamelContext context = getContext();
 
                 try (InputStream is = URIResolver.resolve(context, source)) {
-                    String name = source.getLocation();
-                    name = StringUtils.substringAfter(name, ":");
+                    String name = source.getName();
                     name = StringUtils.removeEnd(name, ".java");
 
-                    if (name.contains("/")) {
-                        name = StringUtils.substringAfterLast(name, "/");
-                    }
+                    // compile the source in memory
+                    String content = IOUtils.toString(is, StandardCharsets.UTF_8);
+                    Reflect compiled = Reflect.compile(name, content);
 
-                    RoutesBuilder builder = Reflect.compile(name, IOUtils.toString(is, StandardCharsets.UTF_8)).create().get();
+                    // create the builder
+                    RoutesBuilder builder = compiled.create().get();
 
                     // Wrap routes builder
                     includeRoutes(builder);

--- a/runtime/camel-k-runtime-jvm/src/test/java/org/apache/camel/k/jvm/RoutesLoadersTest.java
+++ b/runtime/camel-k-runtime-jvm/src/test/java/org/apache/camel/k/jvm/RoutesLoadersTest.java
@@ -89,6 +89,23 @@ public class RoutesLoadersTest {
     }
 
     @Test
+    public void testLoadJavaWithNameOverride() throws Exception {
+        Source source = Source.create("classpath:MyRoutesWithNameOverride.java?name=MyRoutes.java");
+        RoutesLoader loader = RuntimeSupport.loaderFor(new DefaultCamelContext(), source);
+        RouteBuilder builder = loader.load(new InMemoryRegistry(), source);
+
+        assertThat(loader).isInstanceOf(JavaSourceLoader.class);
+        assertThat(builder).isNotNull();
+
+        builder.configure();
+
+        List<RouteDefinition> routes = builder.getRouteCollection().getRoutes();
+        assertThat(routes).hasSize(1);
+        assertThat(routes.get(0).getInputs().get(0).getEndpointUri()).isEqualTo("timer:tick");
+        assertThat(routes.get(0).getOutputs().get(0)).isInstanceOf(ToDefinition.class);
+    }
+
+    @Test
     public void testLoadJavaWithNestedClass() throws Exception {
         Source source = Source.create("classpath:MyRoutesWithNestedClass.java");
         RoutesLoader loader = RuntimeSupport.loaderFor(new DefaultCamelContext(), source);

--- a/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutes.java
+++ b/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutes.java
@@ -1,3 +1,19 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 import org.apache.camel.builder.RouteBuilder;
 
 public class MyRoutes extends RouteBuilder {

--- a/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNameOverride.java
+++ b/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNameOverride.java
@@ -1,3 +1,19 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 import org.apache.camel.builder.RouteBuilder;
 
 public class MyRoutes extends RouteBuilder {

--- a/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNameOverride.java
+++ b/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNameOverride.java
@@ -1,0 +1,9 @@
+import org.apache.camel.builder.RouteBuilder;
+
+public class MyRoutes extends RouteBuilder {
+    @Override
+    public void configure() throws Exception {
+        from("timer:tick")
+                .to("log:info");
+    }
+}

--- a/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNestedClass.java
+++ b/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithNestedClass.java
@@ -1,3 +1,19 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 import org.apache.camel.Exchange;
 import org.apache.camel.Processor;
 import org.apache.camel.builder.RouteBuilder;

--- a/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithRestConfiguration.java
+++ b/runtime/camel-k-runtime-jvm/src/test/resources/MyRoutesWithRestConfiguration.java
@@ -1,3 +1,19 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 import org.apache.camel.builder.RouteBuilder;
 
 public class MyRoutesWithRestConfiguration extends RouteBuilder {

